File: repos/firecrawl/apps/php-sdk/src/Client/FirecrawlHttpClient.php
<?php

declare(strict_types=1);

namespace Firecrawl\Client;

use Firecrawl\Exceptions\AuthenticationException;
use Firecrawl\Exceptions\FirecrawlException;
use Firecrawl\Exceptions\RateLimitException;
use Firecrawl\Version;
use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\ClientInterface;
use GuzzleHttp\Exception\ConnectException;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\RequestOptions;

/**
 * @internal
 */
final class FirecrawlHttpClient
{
    private readonly ClientInterface $httpClient;
    private readonly string $baseUrl;

    public function __construct(
        private readonly string $apiKey,
        string $baseUrl,
        float $timeoutSeconds,
        private readonly int $maxRetries,
        private readonly float $backoffFactor,
        ?ClientInterface $httpClient = null,
    ) {
        $this->baseUrl = rtrim($baseUrl, '/');

        $this->httpClient = $httpClient ?? new GuzzleClient([
            RequestOptions::TIMEOUT => $timeoutSeconds,
            RequestOptions::CONNECT_TIMEOUT => min($timeoutSeconds, 30),
            RequestOptions::HTTP_ERRORS => false,
        ]);
    }

    /**
     * @param array<string, mixed>  $body
     * @param array<string, string> $extraHeaders
     * @return array<string, mixed>
     */
    public function post(string $path, array $body, array $extraHeaders = [], ?float $timeoutSeconds = null): array
    {
        return $this->request('POST', $this->baseUrl . $path, $body, $extraHeaders, timeoutSeconds: $timeoutSeconds);
    }

    /** @return array<string, mixed> */
    public function get(string $path, ?float $timeoutSeconds = null): array
    {
        return $this->request('GET', $this->baseUrl . $path, timeoutSeconds: $timeoutSeconds);
    }

    /** @return array<string, mixed> */
    public function getAbsolute(string $absoluteUrl): array
    {
        return $this->request('GET', $absoluteUrl);
    }

    /** @return array<string, mixed> */
    public function delete(string $path): array
    {
        return $this->request('DELETE', $this->baseUrl . $path);
    }

    /**
     * @param array<string, mixed> $body
     * @return array<string, mixed>
     */
    public function patch(string $path, array $body): array
    {
        return $this->request('PATCH', $this->baseUrl . $path, $body);
    }

    /**
     * Send a POST request with a multipart/form-data body.
     *
     * @param array<string, string> $fields
     * @return array<string, mixed>
     */
    public function postMultipart(
        string $path,
        array $fields,
        string $fileField,
        string $fileName,
        string $fileContent,
        ?string $fileContentType = null,
    ): array {
        $multipart = [];
        foreach ($fields as $name => $value) {
            $multipart[] = [
                'name' => $name,
                'contents' => $value,
            ];
        }

        $filePart = [
            'name' => $fileField,
            'contents' => $fileContent,
            'filename' => $fileName,
        ];
        if ($fileContentType !== null && $fileContentType !== '') {
            $filePart['headers'] = ['Content-Type' => $fileContentType];
        }
        $multipart[] = $filePart;

        return $this->request(
            'POST',
            $this->baseUrl . $path,
            body: [],
            extraHeaders: [],
            multipart: $multipart,
        );
    }

    public function getBaseUrl(): string
    {
        return $this->baseUrl;
    }

    /**
     * @param array<string, mixed>                  $body
     * @param array<string, string>                 $extraHeaders
     * @param list<array<string, mixed>>|null       $multipart
     * @return array<string, mixed>
     */
    private function request(
        string $method,
        string $url,
        array $body = [],
        array $extraHeaders = [],
        ?array $multipart = null,
        ?float $timeoutSeconds = null,
    ): array {
        $defaultHeaders = [
            'Accept' => 'application/json',
            'User-Agent' => 'firecrawl-php/' . Version::SDK_VERSION,
        ];
        // Omit the Authorization header entirely when no key is set so that
        // scrape/search/interact can use the keyless free tier.
        if ($this->apiKey !== '') {
            $defaultHeaders['Authorization'] = 'Bearer ' . $this->apiKey;
        }

        if ($multipart === null) {
            $defaultHeaders['Content-Type'] = 'application/json';
        }

        $headers = array_merge($defaultHeaders, $extraHeaders);

        $options = [
            RequestOptions::HEADERS => $headers,
            RequestOptions::HTTP_ERRORS => false,
        ];

        if ($timeoutSeconds !== null) {
            $options[RequestOptions::TIMEOUT] = $timeoutSeconds;
        }

        if ($multipart !== null) {
            $options[RequestOptions::MULTIPART] = $multipart;
        } elseif (($method === 'POST' || $method === 'PATCH') && $body !== []) {
            $options[RequestOptions::JSON] = $body;
        }

        $attempt = 0;

        while (true) {
            try {
                $response = $this->httpClient->request($method, $url, $options);
                $statusCode = $response->getStatusCode();
                $responseBody = (string) $response->getBody();

                if ($statusCode >= 200 && $statusCode < 300) {
                    if ($responseBody === '' || $responseBody === '{}') {
                        return [];
                    }
                    /** @var array<string, mixed> */
                    return json_decode($responseBody, true, 512, JSON_THROW_ON_ERROR);
                }

                $errorMessage = $this->extractErrorMessage($responseBody, $statusCode);
                $errorCode = $this->extractErrorCode($responseBody);

                // Non-retryable client errors
                if ($statusCode === 401) {
                    throw new AuthenticationException($errorMessage, $errorCode);
                }

                if ($statusCode === 429) {
                    throw new RateLimitException($errorMessage, $errorCode);
                }

                if ($statusCode >= 400 && $statusCode < 500 && $statusCode !== 408 && $statusCode !== 409) {
                    throw new FirecrawlException($errorMessage, $statusCode, $errorCode);
                }

                // Retryable errors: 408, 409, 502, 5xx
                if ($attempt < $this->maxRetries) {
                    $attempt++;
                    $this->sleepWithBackoff($attempt);
                    continue;
                }

                throw new FirecrawlException($errorMessage, $statusCode, $errorCode);
            } catch (FirecrawlException $e) {
                throw $e;
            } catch (ConnectException $e) {
                if ($attempt < $this->maxRetries) {
                    $attempt++;
                    $this->sleepWithBackoff($attempt);
                    continue;
                }
                throw new FirecrawlException('Connection failed: ' . $e->getMessage(), previous: $e);
            } catch (RequestException $e) {
                if ($attempt < $this->maxRetries) {
                    $attempt++;
                    $this->sleepWithBackoff($attempt);
                    continue;
                }
                throw new FirecrawlException('Request failed: ' . $e->getMessage(), previous: $e);
            } catch (\JsonException $e) {
                throw new FirecrawlException('Failed to parse API response: ' . $e->getMessage(), previous: $e);
            }
        }
    }

    private function extractErrorMessage(string $body, int $statusCode): string
    {
        try {
            /** @var array<string, mixed> $parsed */
            $parsed = json_decode($body, true, 512, JSON_THROW_ON_ERROR);

            if (isset($parsed['error']) && is_string($parsed['error'])) {
                return $parsed['error'];
            }

            if (isset($parsed['message']) && is_string($parsed['message'])) {
                return $parsed['message'];
            }
        } catch (\JsonException) {
            // ignore
        }

        return "HTTP {$statusCode} error";
    }

    private function extractErrorCode(string $body): ?string
    {
        try {
            /** @var array<string, mixed> $parsed */
            $parsed = json_decode($body, true, 512, JSON_THROW_ON_ERROR);

            if (isset($parsed['code'])) {
                return (string) $parsed['code'];
            }
        } catch (\JsonException) {
            // ignore
        }

        return null;
    }

    private function sleepWithBackoff(int $attempt): void
    {
        $delayMs = (int) ($this->backoffFactor * 1000 * pow(2, $attempt - 1));
        usleep($delayMs * 1000);
    }
}

File: repos/firecrawl/apps/java-sdk/src/main/java/com/firecrawl/client/FirecrawlHttpClient.java
package com.firecrawl.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jdk8.Jdk8Module;
import com.firecrawl.errors.AuthenticationException;
import com.firecrawl.errors.FirecrawlException;
import com.firecrawl.errors.RateLimitException;
import okhttp3.*;

import java.io.IOException;
import java.util.Collections;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Internal HTTP client for making authenticated requests to the Firecrawl API.
 * Handles retry logic with exponential backoff.
 */
class FirecrawlHttpClient {

    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private final OkHttpClient httpClient;
    private final String apiKey;
    private final String baseUrl;
    private final int maxRetries;
    private final double backoffFactor;
    final ObjectMapper objectMapper;

    FirecrawlHttpClient(String apiKey, String baseUrl, long timeoutMs, int maxRetries, double backoffFactor) {
        this(apiKey, baseUrl, timeoutMs, maxRetries, backoffFactor, null);
    }

    FirecrawlHttpClient(String apiKey, String baseUrl, long timeoutMs, int maxRetries, double backoffFactor,
                         OkHttpClient httpClient) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.maxRetries = maxRetries;
        this.backoffFactor = backoffFactor;

        if (httpClient != null) {
            this.httpClient = httpClient;
        } else {
            this.httpClient = new OkHttpClient.Builder()
                    .connectTimeout(timeoutMs, TimeUnit.MILLISECONDS)
                    .readTimeout(timeoutMs, TimeUnit.MILLISECONDS)
                    .writeTimeout(timeoutMs, TimeUnit.MILLISECONDS)
                    .build();
        }

        this.objectMapper = new ObjectMapper()
                .registerModule(new Jdk8Module())
                .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    /**
     * Adds the Authorization header only when an API key is configured. Omitting
     * it entirely (rather than sending an empty Bearer) lets scrape/search/interact
     * use the keyless free tier.
     */
    private void applyAuth(Request.Builder builder) {
        if (apiKey != null && !apiKey.isBlank()) {
            builder.header("Authorization", "Bearer " + apiKey);
        }
    }

    /**
     * Sends a POST request with JSON body.
     */
    <T> T post(String path, Object body, Class<T> responseType) {
        return post(path, body, responseType, Collections.emptyMap());
    }

    /**
     * Sends a POST request with JSON body and extra headers.
     */
    <T> T post(String path, Object body, Class<T> responseType, Map<String, String> extraHeaders) {
        String url = baseUrl + path;
        String json;
        try {
            json = objectMapper.writeValueAsString(body);
        } catch (JsonProcessingException e) {
            throw new FirecrawlException("Failed to serialize request body", e);
        }
        RequestBody requestBody = RequestBody.create(json, JSON);
        Request.Builder builder = new Request.Builder()
                .url(url)
                .header("Content-Type", "application/json")
                .post(requestBody);
        applyAuth(builder);
        for (Map.Entry<String, String> entry : extraHeaders.entrySet()) {
            builder.header(entry.getKey(), entry.getValue());
        }
        Request request = builder.build();
        return executeWithRetry(request, responseType);
    }

    /**
     * Sends a PATCH request with JSON body.
     */
    <T> T patch(String path, Object body, Class<T> responseType) {
        String url = baseUrl + path;
        String json;
        try {
            json = objectMapper.writeValueAsString(body);
        } catch (JsonProcessingException e) {
            throw new FirecrawlException("Failed to serialize request body", e);
        }
        RequestBody requestBody = RequestBody.create(json, JSON);
        Request.Builder builder = new Request.Builder()
                .url(url)
                .header("Content-Type", "application/json")
                .patch(requestBody);
        applyAuth(builder);
        Request request = builder.build();
        return executeWithRetry(request, responseType);
    }

    /**
     * Sends a POST multipart/form-data request.
     */
    <T> T postMultipart(
            String path,
            Map<String, String> fields,
            String fileFieldName,
            byte[] fileContent,
            String filename,
            String contentType,
            Class<T> responseType
    ) {
        String url = baseUrl + path;
        MultipartBody.Builder multipart = new MultipartBody.Builder()
                .setType(MultipartBody.FORM);

        for (Map.Entry<String, String> entry : fields.entrySet()) {
            multipart.addFormDataPart(entry.getKey(), entry.getValue());
        }

        MediaType mediaType;
        if (contentType != null && !contentType.isBlank()) {
            try {
                mediaType = MediaType.get(contentType);
            } catch (IllegalArgumentException ignored) {
                mediaType = MediaType.get("application/octet-stream");
            }
        } else {
            mediaType = MediaType.get("application/octet-stream");
        }
        RequestBody fileBody = RequestBody.create(fileContent, mediaType);
        multipart.addFormDataPart(fileFieldName, filename, fileBody);

        Request.Builder builder = new Request.Builder()
                .url(url)
                .post(multipart.build());
        applyAuth(builder);
        Request request = builder.build();

        return executeWithRetry(request, responseType);
    }

    /**
     * Sends a GET request.
     */
    <T> T get(String path, Class<T> responseType) {
        String url = baseUrl + path;
        Request.Builder builder = new Request.Builder()
                .url(url)
                .get();
        applyAuth(builder);
        Request request = builder.build();
        return executeWithRetry(request, responseType);
    }

    /**
     * Sends a GET request with full URL (for following next-page cursors).
     */
    <T> T getAbsolute(String absoluteUrl, Class<T> responseType) {
        Request.Builder builder = new Request.Builder()
                .url(absoluteUrl)
                .get();
        applyAuth(builder);
        Request request = builder.build();
        return executeWithRetry(request, responseType);
    }

    /**
     * Sends a DELETE request.
     */
    <T> T delete(String path, Class<T> responseType) {
        String url = baseUrl + path;
        Request.Builder builder = new Request.Builder()
                .url(url)
                .delete();
        applyAuth(builder);
        Request request = builder.build();
        return executeWithRetry(request, responseType);
    }

    /**
     * Sends a raw GET request and returns the response body as a parsed Map.
     */
    @SuppressWarnings("unchecked")
    Map<String, Object> getRaw(String path) {
        return get(path, Map.class);
    }

    private <T> T executeWithRetry(Request request, Class<T> responseType) {
        int attempt = 0;
        while (true) {
            try {
                try (Response response = httpClient.newCall(request).execute()) {
                    ResponseBody responseBody = response.body();
                    String bodyStr = responseBody != null ? responseBody.string() : "";

                    if (response.isSuccessful()) {
                        if (responseType == Void.class || responseType == void.class) {
                            return null;
                        }
                        return objectMapper.readValue(bodyStr, responseType);
                    }

                    int code = response.code();

                    // Parse error details from response
                    String errorMessage = extractErrorMessage(bodyStr, code);
                    String errorCode = extractErrorCode(bodyStr);

                    // Non-retryable client errors
                    if (code == 401) {
                        throw new AuthenticationException(errorMessage, errorCode, null);
                    }
                    if (code == 429) {
                        throw new RateLimitException(errorMessage, errorCode, null);
                    }
                    if (code >= 400 && code < 500 && code != 408 && code != 409) {
                        throw new FirecrawlException(errorMessage, code, errorCode, null);
                    }

                    // Retryable errors: 408, 409, 502, 5xx
                    if (attempt < maxRetries) {
                        attempt++;
                        sleepWithBackoff(attempt);
                        continue;
                    }

                    throw new FirecrawlException(errorMessage, code, errorCode, null);
                }
            } catch (FirecrawlException e) {
                throw e;
            } catch (IOException e) {
                if (attempt < maxRetries) {
                    attempt++;
                    sleepWithBackoff(attempt);
                    continue;
                }
                throw new FirecrawlException("Request failed: " + e.getMessage(), e);
            }
        }
    }

    @SuppressWarnings("unchecked")
    private String extractErrorMessage(String body, int statusCode) {
        try {
            Map<String, Object> parsed = objectMapper.readValue(body, Map.class);
            if (parsed.containsKey("error")) {
                return String.valueOf(parsed.get("error"));
            }
            if (parsed.containsKey("message")) {
                return String.valueOf(parsed.get("message"));
            }
        } catch (Exception ignored) {
        }
        return "HTTP " + statusCode + " error";
    }

    @SuppressWarnings("unchecked")
    private String extractErrorCode(String body) {
        try {
            Map<String, Object> parsed = objectMapper.readValue(body, Map.class);
            Object code = parsed.get("code");
            return code != null ? String.valueOf(code) : null;
        } catch (Exception ignored) {
        }
        return null;
    }

    private void sleepWithBackoff(int attempt) {
        long delayMs = (long) (backoffFactor * 1000 * Math.pow(2, attempt - 1));
        try {
            Thread.sleep(delayMs);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new FirecrawlException("Request interrupted during retry backoff", e);
        }
    }
}
